books/RayTracingTheRestOfYourLife.html

```diff
@@ -423,184 +571,504 @@
+
+Our, new, crude approximation for $\frac{r}{2}$ is nonuniform (but only just):
+
+    ![Figure [approx-f]: A nonuniform distribution for r/2](../images/fig-3.06-nonuniform-dist.jpg)
+
+We had the analytical solution to the integration above, so we could very very easily solve for the
+50% value. We could also solve for this experimentally. If we had a function that couldn't solve
+for the integration for--or didn't _want_ to solve the integration for--we could get an experimental
+result close to the real value. Let's take the function:
+
+    $$ p(x) = e^{\frac{-x}{2 \pi}} sin^2(x) $$
+
+Which looks a little something like this:
+
+    ![Figure [exp-sin2]: A function that we don't want to solve analytically](../images/fig-3.07-exp-sin2.jpg)
+
+From $0$ to $2\pi$ as example:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    #include "rtweekend.h"
+
+    #include <algorithm>
+    #include <vector>
+    #include <iostream>
+    #include <iomanip>
+    #include <math.h>
+    #include <cmath>
+    #include <stdlib.h>
+
+    int main() {
+        int N = 10000;
+        double sum = 0.0;
+        // Get all of our samples
+        // Get the area under the curve
+        std::vector<std::pair<double, double>> samples;
+        for (int i = 0; i < N; i++) {
+            auto x = random_double(0,2 * pi);
+            auto p_x = exp(-x/ (2 * pi)) * sin(x) * sin(x);
+            sum += p_x;
+            for(int s = 0; s < samples.size(); s++)
+            {
+                // sort by x
+                if(x > samples[s].first) {
+                    samples.insert(samples.begin() + s, std::pair<double, double>(x, p_x));
+                    break;
+                }
+            }
+            samples.push_back(std::pair<double, double>(x, p_x));
+        }
+
+        // Find out the sample at which we have half of our area
+        double half_sum = sum / 2.0;
+        double halfway_point;
+        double accum = 0.0;
+        for (int i = 0; i < N; i++){
+            accum += samples[i].second;
+            if (accum >= half_sum){
+                halfway_point = samples[i].first;
+                break;
+            }
+        }
+
+        std::cout << std::fixed << std::setprecision(12);
+        std::cout << "Average = " << sum / N << '\n';
+        std::cout << "Area under curve = " << 2 * pi * sum / N << '\n';
+        std::cout << "Halfway = " << halfway_point << '\n';
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [est-halfway]: <kbd>[estimate_halfway.cc]</kbd> Estimating the 50% point of a function]
+
+This code snippet isn't too different from what we had before. We're still solving for the sum
+over an interval (0 to $2\pi$). Only this time, we're also storing and sorting all of our samples by
+their input and output. We then use this to determine how many of the original samples sum up to
```

Comment:
“We then use this to determine the point at which they subtotal up to half of the sum ...”